File: Airflow_orchestrator/Airflow_Setup/dags/scripts/wordcount.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import explode, split, col, desc

import sys

def main():
    if len(sys.argv) < 3:
        print("Usage: wordcount.py <input_file> <output_directory>", file=sys.stderr)
        sys.exit(1)

    input_file = sys.argv[1]
    output_dir = sys.argv[2]

    spark = SparkSession.builder.appName("WordCount").getOrCreate()
    spark.conf.set("spark.custom.jobName", "WordCount")

    # ✅ Read text file and create temp view
    lines = spark.read.text(input_file)
    lines.createOrReplaceTempView("lines")

    # ✅ SQL query equivalent
    words = spark.sql("""
        SELECT explode(split(value, '\\s+')) AS word
        FROM lines
    """).filter(col("word") != "")

    # ✅ Group by word and count
    counts = words.groupBy("word").count().orderBy(desc("count"))

    # ✅ Write result to output
    counts.write.format("csv").save(output_dir)

    spark.stop()

if __name__ == "__main__":
    main()

File: transformations/wordcount/src/main/java/sdu/masters/WordCount.java
package sdu.masters;

import org.apache.spark.sql.*;
import org.apache.spark.api.java.function.FilterFunction;


public class WordCount {
    public static void main(String[] args) {
        if (args.length < 2) {
            System.err.println("Usage: WordCount <input_file> <output_directory>");
            System.exit(1);
        }

        SparkSession spark = SparkSession.builder()
                .appName("WordCount")
                .getOrCreate();
        spark.conf().set("spark.custom.jobName", "wordcount");
        //spark.sparkContext().addSparkListener(new GlobalMetadataListener());
        //spark.listenerManager().register(new QueryMetadataListener());

        // ✅ Read as DataFrame and create temp view
        Dataset<String> lines = spark.read().textFile(args[0]);
        lines.createOrReplaceTempView("lines");

        // ✅ SQL that triggers query execution
        Dataset<Row> words = spark.sql("SELECT explode(split(value, '\\\\s+')) AS word FROM lines")
                .filter((FilterFunction<Row>) row -> !row.getString(0).isEmpty());

        Dataset<Row> counts = words.groupBy("word").count().orderBy(functions.desc("count"));

        counts.write().format("txt").save(args[1]);

        spark.stop();
    }
}
